File: @mzm/log/render.ts
import {colors} from '@cliffy/ansi/colors'
import type {LogLine} from './types.ts'

const ERROR_REGEX = /err(?:or)?|crit(?:ical)?|fatal|severe|emerg(?:ENCY)?/i
const WARN_REGEX = /warn(?:ing)?/i
const INFO_REGEX = /info/i
const DEBUG_REGEX = /debug/i
const TRACE_REGEX = /trace/i

const THEME = {
  warn: (value: string) => {
    return colors.black.bgYellow(value)
  },
  error: (value: string) => {
    return colors.white.bgRed(value)
  },
  info: (value: string) => {
    return colors.green.bgBlack(value)
  },
  trace: (value: string) => {
    return colors.blue.bgBlack(value)
  },
  debug: (value: string) => {
    return colors.dim.bgBlack(value)
  },
}

export function pprint(line: LogLine, ) {
  if (!line._host) return ''
  const timestamp = new Date(line._ts).toString()
  const host = line._host
  const app = line._app
  const message = line.message || line._line
  const level = line.level

  const date_string = timestamp.substring(4, 11) + timestamp.substring(16, 24)
  return [
    colors.gray(date_string),
    colors.red(host),
    colors.blue(app),
    colorize(level),
    text(message, level),
  ].join(' ')
}

function colorize(level: string = 'info') {
  if (ERROR_REGEX.test(level)) return THEME.error(` ${level} `)
  if (WARN_REGEX.test(level)) return THEME.warn(` ${level} `)
  if (INFO_REGEX.test(level)) return THEME.info(` ${level} `)
  if (DEBUG_REGEX.test(level)) return THEME.debug(` ${level} `)
  if (TRACE_REGEX.test(level)) return THEME.trace(` ${level} `)
  return level
}

function text(message: string, level: string) {
  if (level == 'debug') return colors.gray(message)
  if (level == 'trace') return colors.gray(message)
  return message
}

File: commands/log/render.go
package log

import (
	JSON "encoding/json"
	"fmt"
	"log"
	"regexp"
	"strconv"
	"strings"
	"time"

	style "github.com/phoenix-tui/phoenix/style"
)

var errorRegex = regexp.MustCompile(`(?i)err(?:or)?|crit(?:ical)?|fatal|severe|emerg(?:ency)?`)
var warnRegex = regexp.MustCompile(`(?i)warn(?:ing)?`)
var infoRegex = regexp.MustCompile(`(?i)info`)
var debugRegex = regexp.MustCompile(`(?i)debug`)
var traceRegex = regexp.MustCompile(`(?i)trace`)
var THEME = make(map[string]style.Style)
var dimStyle style.Style
var redStyle style.Style
var blueStyle style.Style
var grayStyle style.Style
var levelMap LevelMap

const TIME_FORMAT string = "Jan 02 15:04:05"
const LVL_TRACE = "trace"
const LVL_DEBUG = "debug"
const SPACE = " "

var center style.Alignment = style.NewAlignment(style.AlignCenter, style.AlignMiddle)
var defaultStyle style.Style = style.New().
	Foreground(style.RGB(108, 108, 108)).
	Padding(style.NewPadding(0, 1, 0, 1)).
	Width(10).
	MaxWidth(10).
	Align(center)

type StyleValue struct {
	Pattern *regexp.Regexp
	Style   style.Style
}

type LevelMap = []StyleValue

type Message struct {
	Host      string `json:"_host"`
	Timestamp uint64 `json:"_ts"`
	App       string `json:"_app"`
	Msg       string `json:"message,omitempty"`
	Line      string `json:"_line,omitempty"`
	Level     any    `json:"level"`
}

// levelToString safely converts a level field of any type to a string
func levelToString(level any) string {
	switch v := level.(type) {
	case string:
		return v
	case int:
		return strconv.Itoa(v)
	case int64:
		return strconv.FormatInt(v, 10)
	case float64:
		// JSON numbers are unmarshaled as float64 by default
		return strconv.FormatFloat(v, 'f', -1, 64)
	case fmt.Stringer:
		return v.String()
	default:
		// Fallback to sprintf for any other type
		return fmt.Sprintf("%v", v)
	}
}

func colorize(level string) string {
	for _, def := range levelMap {
		match := def.Pattern.MatchString(level)
		if match == true {
			return style.Render(def.Style, level)
		}
	}
	return style.Render(defaultStyle, level)
}

func text(message string, level string) string {
	if level == LVL_DEBUG || level == LVL_TRACE {
		return style.Render(dimStyle, message)
	}
	return message
}

// TODO(esatterwhite): replace struct based json codec
// with github.com/buger/jsonparser implementation
// to optimize performance and mimimize memory usage
func pprint(line Message, plain bool) string {
	var message string
	if line.Msg != "" {
		message = line.Msg
	} else {
		message = line.Line
	}

	if plain {
		output, err := JSON.Marshal(line)
		if err != nil {
			log.Fatalf("Unable to format log data for display: %s", err)
		}

		return string(output)

	}
	t := time.UnixMilli(int64(line.Timestamp))
	levelStr := levelToString(line.Level)
	return strings.Join([]string{
		style.Render(grayStyle, t.Format(TIME_FORMAT)),
		style.Render(redStyle, line.Host),
		style.Render(blueStyle, line.App),
		colorize(strings.ToLower(levelStr)),
		text(message, levelStr),
	}, SPACE)
}

func init() {
	colorRed := style.RGB(255, 95, 95)
	dimStyle = style.New().Foreground(style.Gray)
	redStyle = style.New().Foreground(colorRed)
	blueStyle = style.New().Foreground(style.RGB(0, 175, 215))
	grayStyle = style.New().Foreground(style.Color256(24))

	errorStyle := style.New().
		Foreground(style.White).
		Background(colorRed).
		Padding(style.NewPadding(0, 1, 0, 1)).
		Width(10).
		MaxWidth(10).
		Align(center)

	infoStyle := style.New().
		Foreground(style.Color256(155)).
		Background(style.RGB(88, 88, 88)).
		Padding(style.NewPadding(0, 1, 0, 1)).
		Width(10).
		MaxWidth(10).
		Align(center)

	traceStyle := style.New().
		Foreground(style.RGB(0, 95, 215)).
		Background(style.RGB(99, 99, 99)).
		Padding(style.NewPadding(0, 1, 0, 1)).
		Width(10).
		MaxWidth(10).
		Align(center)

	var debugStyle style.Style = defaultStyle

	warnStyle := style.New().
		Foreground(style.Black).
		Background(style.RGB(255, 215, 0)).
		Padding(style.NewPadding(0, 1, 0, 1)).
		Width(10).
		MaxWidth(10).
		Align(center)

	levelMap = LevelMap{
		StyleValue{
			Pattern: errorRegex,
			Style:   errorStyle,
		},
		StyleValue{
			Pattern: infoRegex,
			Style:   infoStyle,
		},
		StyleValue{
			Pattern: debugRegex,
			Style:   debugStyle,
		},
		StyleValue{
			Pattern: traceRegex,
			Style:   traceStyle,
		},
		StyleValue{
			Pattern: warnRegex,
			Style:   warnStyle,
		},
	}
}
